src/herder/TxSetFrame.cpp

```diff
@@ -347,207 +268,416 @@ TxSetFrame::makeFromTransactions(TxPhases const& txPhases, Application& app,
+    }
+    invalidTxs = invalid[0];
+    return res;
 }
+#endif
 
-TxSetFrameConstPtr
-TxSetFrame::makeFromWire(Application& app,
-                         GeneralizedTransactionSet const& xdrTxSet)
+ResolvedTxSetFrame const*
+TxSetFrame::resolve(Application& app, AbstractLedgerTxn& ltx) const
 {
+    if (mResolvedTxSet)
+    {
+        return mResolvedTxSet->get();
+    }
     ZoneScoped;
-    auto hash = xdrSha256(xdrTxSet);
-    size_t encodedSize = xdr::xdr_argpack_size(xdrTxSet);
-    if (!validateTxSetXDRStructure(xdrTxSet))
+    releaseAssert(previousLedgerHash() ==
+                  app.getLedgerManager().getLastClosedLedgerHeader().hash);
+
+    if (isGeneralizedTxSet())
     {
-        return std::make_shared<InvalidTxSetFrame const>(xdrTxSet, hash,
-                                                         encodedSize);
+        auto const& xdrTxSet = std::get<GeneralizedTransactionSet>(mXDRTxSet);
+        if (!validateTxSetXDRStructure(xdrTxSet))
+        {
+            CLOG_DEBUG(Herder,
+                       "Got bad generalized txSet with invalid XDR structure");
+            return nullptr;
+        }
+        auto const& phases = xdrTxSet.v1TxSet().phases;
+        TxSetFrame::TxPhases defaultPhases;
+        defaultPhases.resize(phases.size());
+
+        std::unique_ptr<ResolvedTxSetFrame> txSet(new ResolvedTxSetFrame(
+            true, xdrTxSet.v1TxSet().previousLedgerHash, defaultPhases));
+        // Mark fees as already computed as we read them from the XDR.
+        for (int i = 0; i < txSet->mFeesComputed.size(); i++)
+        {
+            txSet->mFeesComputed[i] = true;
+        }
+
+        releaseAssert(phases.size() <=
+                      static_cast<size_t>(TxSetFrame::Phase::PHASE_COUNT));
+        for (auto phaseId = 0; phaseId < phases.size(); phaseId++)
+        {
+            auto const& phase = phases[phaseId];
+            auto const& components = phase.v0Components();
+            for (auto const& component : components)
+            {
+                switch (component.type())
+                {
+                case TXSET_COMP_TXS_MAYBE_DISCOUNTED_FEE:
+                    std::optional<int64_t> baseFee;
+                    if (component.txsMaybeDiscountedFee().baseFee)
+                    {
+                        baseFee = *component.txsMaybeDiscountedFee().baseFee;
+                    }
+                    if (!txSet->addTxsFromXdr(
+                            app, ltx, component.txsMaybeDiscountedFee().txs,
+                            true, baseFee,
+                            static_cast<TxSetFrame::Phase>(phaseId)))
+                    {
+                        CLOG_DEBUG(Herder,
+                                   "Got bad txSet: transactions are not "
+                                   "ordered correctly or contain invalid phase "
+                                   "transactions");
+                        mResolvedTxSet = std::make_optional(nullptr);
```

Comment:
Well, I'm not 100% sure about validity caching (because currently this state only 'virtually' exist for any caller of `checkValid`). I'm not necessarily against adding the states explicitly,, but I'm also not a fan of adding dead code - so if we add states, who'll use them? FWIW now these just describe usage invariants (so that if `resolve` returns non-null, then it'll always return non-null etc.)

Reply:
mm, I didn't think it'd be dead code. The idea was to reason about the tx set based on explicit states, this way externally we can distinguish between unresolved and resolved but invalid, for example (also we won't need an optional in the implementation). It could also be helpful in tests. But it's just a suggestion, no need to do it if you feel strongly.

Reply:
This is no longer relevant, resolving.